### PR TITLE
batch norm support in nvfuser

### PR DESCRIPTION
**Pull request by jjsjann123**
Continuing the work from Ryan's previous PR #657, this PR add batch norm support in nvfuser.

A few topics worth mentioning:

1. inplace update on input
BatchNorm has in-place update on running_mean/running_var in its forward path during training. This is currently supported by a special output to input aliasing, where instead of allocate fresh tensor, we alias output to input to mimic in-place update. This is by no mean considered as aliasing support, but merely a temporary WAR to allow BatchNorm support at this time.
Assume we have an input Tensor0 that is to be updated by the kernel, it currently works as:
```
fusion.addInput(Tensor0)
... # compute the output value to TensorAlias0.
fusion.addOutput(TensorAlias0)
fusion.aliasOutputToInput(TensorAlias0, Tensor0)
```
This means that, instead of allocating TensorAlias0 as a fresh tensor to store output, we'll just re-use Tensor0 for both the input and the output. So effectively we are aliasing the two tensors and the ended behavior is just like in-place update.

2. PyTorch BatchNorm API
Autodiff currently uses `_batch_norm_impl_index` and `_batch_norm_impl_index_backward` which has some rough corners on the API, including 1. a byte Tensor as input and outputs; 2. a scalar integer output. Those are not currently supported by nvfuser, so we ended up adding a preprocessing pass to alter the graph.

- [x] Added Output to Input aliasing in integration
- [x] Re-enabled batch norm in autodiff (There's a separate upstream PR for this https://github.com/pytorch/pytorch/pull/57321)
- [x] Added batch norm parsing rule
- [x] Added necessary pre-processing on PyTorch JIT IR to re-route certain BN inputs/outputs
- [x] Added python test to verify nvfuser BN compatibility with various running mode (affine/track_running_stats/training)

### CODE REVIEW

**Comment by csarofeen** [comment]
@jjsjann123 is this PR still active?

**Comment by jjsjann123** [comment]
Yes, it's active. I just need a working kernel so I can test/validate the integration code.
Haven't got it cleaned up yet, will do and then remove the [DO NOT REVIEW] tag.

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -2120,6 +2125,124 @@ def t(x):
...
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_batch_norm_impl_index_correctness(self):
+        batch = [2, 7, 16]
+        channels = [4, 19, 32, 89]
+        hw = [1, 8, 17, 32]
+
+        # failing sizes (2, 1, 1, 1)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:2235]: I need to double check this, but let's worry about it later.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 2]
```diff
@@ -422,11 +422,16 @@ FusionExecutor::GlobalBuffers FusionExecutor::allocGlobalVals(
 }
 
 std::vector<at::Tensor> FusionExecutor::allocOutputs(
-    kir::ExpressionEvaluator& expr_eval) {
+    kir::ExpressionEvaluator& expr_eval,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/executor.cpp:425]: The skip_io part for aliasing is quite hacky. I think it should work but might get complicated with segmentation.
Need to clean it up later and sort thing straight with segmentation.
Meanwhile, cc'ing @shmsong for a brief review on this.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 3]
```diff
@@ -1470,35 +1612,38 @@ void CudaFuseGraph(std::shared_ptr<Graph>& graph) {
...
-  GRAPH_DUMP("After Fusion: ", graph);
+  GRAPH_DEBUG("After Fusion: ", *graph);
 
   // guard input types as well as conditional constants from
   // aten::profile_ivalue
   guardFusionGroups(graph->block());
-  GRAPH_DUMP("After Guard Fusion: ", graph);
+  GRAPH_DEBUG("After Guard Fusion: ", *graph);
+
+  alterBatchNormImpls(graph->block());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1711]: these are just necessary graph mutation to make PyTorch BN graph to be compatible with nvfuser.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -613,116 +616,359 @@ class IrParser {
           });
     }
 
+    {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:619]: @rdspring1 Would you mind to take a quick look here since I copied most of this from your old PR.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:619]: note, this should have been cleaned up further. We briefly talked about putting some higher level pattern in nn_arith.cpp. We can revisit this then.

File: torch/csrc/jit/codegen/cuda/partition.cpp [hunk 5]
```diff
@@ -74,8 +106,8 @@ inline bool isFusibleNode(const Node* node) {
   // Check we have a parsing rule
   bool isFusible = isNodeParsible(node);
   // Check if we have a tensor type it's one we support
-  isFusible = isFusible && allCompatableTensorTypes(node->inputs());
-  isFusible = isFusible && allCompatableTensorTypes(node->outputs());
+  isFusible = isFusible && checkInputTensorTypes(node);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/partition.cpp:109]: Have to make these changes so we won't exclude BN for its input/output tensor scalar types.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 4 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp:619]: I started creating a high level API for softmax, layer_norm, batch_norm, etc.

File: torch/csrc/jit/codegen/cuda/executor.h [hunk 6]
```diff
@@ -145,7 +146,9 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
   // not initialized, while the second vector contains zero-initiliazed tensors
   GlobalBuffers allocGlobalVals(kir::ExpressionEvaluator& expr_eval);
 
-  std::vector<at::Tensor> allocOutputs(kir::ExpressionEvaluator& expr_eval);
+  std::vector<at::Tensor> allocOutputs(
+      kir::ExpressionEvaluator& expr_eval,
+      int skip_io_aliases = 0);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/executor.h]: Can you please add a comment on `skip_io_aliases`?

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp:425]: I'm trying to make sure I understand this correctly. So `Fusion::aliasOutputToInput` will be the user/parser facing interface that declares an alias and `skip_io_aliases` assumes the aliased outputs will always be the first few ones declared. Is that correct?

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp:425]: IIUC, we are only using aliases for `batchnorm` but how general do we want to fuse it? Essentially this is breaking the SSA of fusion IR and I'd expect a few places to need updating. I will try to figure out a plan to make the segmenter support this correctly in edge cases.

File: aten/src/ATen/native/Normalization.cpp [hunk 7]
```diff
@@ -498,7 +498,9 @@ std::tuple<Tensor, Tensor, Tensor> _batch_norm_impl_index_backward(
   const Tensor& save_mean = c10::value_or_else(save_mean_opt, [] {return Tensor();});
   const Tensor& save_var_transform = c10::value_or_else(save_var_transform_opt, [] {return Tensor();});
 
-  if (impl_index == 0) {
+  // backward in inference mode is not supported in cudnn, fallback to native
+  // TODO: verify the same thing in miopen
```
Comment (csarofeen) [inline aten/src/ATen/native/Normalization.cpp]: Looks like you checked this.

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (csarofeen) [reply test/test_jit_cuda_fuser.py:2235]: Do you have an issue to track it?

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:425]: We're only doing this with BN. I discussed with Jie we can look case by case but the plan isn't to do/use this anywhere else.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 8]
```diff
@@ -479,13 +486,20 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
...
       // take the short-cut for launch if we see a recorded input set again
       launch_params = executor_entry->launch_params;
-      for (size_t i = 0; i < executor_entry->output_sizes.size(); i++) {
-        allocated_outputs.push_back(at::native::empty_cuda(
-            executor_entry->output_sizes[i],
-            executor_entry->output_types[i],
-            c10::nullopt,
-            options_.device,
-            c10::nullopt));
+      if (outputs.empty() || outputs.size() != fusion_.outputs().size()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: Doesn't the rhs of the `||` cover the lhs case?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: I'm confused why we set `allocated_outputs` driectly to `outputs` then have this second part of the conditional. Can you explain the logic here?

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 9]
```diff
@@ -522,9 +536,20 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
     executor_utils::validateVectorizedTensors(
         &fusion_, inputs, outputs, lowered_, expr_eval);
 
+    auto alias_indices = fusion_.getInputAliasIndices();
+
+    // TODO: Update this as well;
     if (outputs.empty() || outputs.size() != fusion_.outputs().size()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: So we just ignore the provided outputs here?

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 10]
```diff
@@ -601,6 +612,55 @@ std::vector<Val*> Fusion::getTerminatingOutputs() {
   return terminating_outputs;
 }
 
+void Fusion::aliasOutputToInput(Val* output, Val* input) {
+  TORCH_INTERNAL_ASSERT(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:617]: I think the check here should be if it has the output, then it must have the input. Not that they should both always be here. Also should add a comment that this is due to segmentation where we may use the input, and we could alias it, but if we don't produce the output then aliasing doesn't have to happen.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:617]: We should also clean up the returned alias if entries aren't in both input and output for a given alias.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:617]: Since remove/add/output/input can be called multiple times on a fusion, we should validate at lowering that if an output in the alias map is there, then the corresponding input must be registered.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 11]
```diff
@@ -1347,8 +1357,140 @@ void guardFusionGroups(Block* block) {
     //         c. restore conditional constant to non-constant for fallback
     guardFusionGroup(fusion);
   }
+}
+
+void alterBatchNormImplIndex(Node* node) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1425]: Comment this function please.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 12]
```diff
@@ -613,116 +616,359 @@ class IrParser {
...
+        auto ptr_op = getOperatorForLiteral(signature);
+        registerParseRule(
+            ptr_op,
+            [](const Node* node,
+               std::unordered_map<size_t, CgValue>& value_map) -> void {
+              auto input =
+                  value_map[node->input(0)->unique()]->as<TensorView>();
+
+              // TODO: it feels quite sketchy to modify fusion from parser
+              auto fusion = FusionGuard::getCurFusion();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:634]: It actually seems okay in this instance, but you can grab the fusion from any value simply with `val->fusion()`. In order to do all the parsing below though you need the fusion active anyways.

**Review by csarofeen — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 8 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: Good catch. I should have removed the lhs of `||`.

In terms of the check, for some reason the function `runFusion` allows user to pass in output instead of allocating them inside this function. Those are used by some tests.
So there's no guarantee that provided output would actually work. Hence the check. There was a bug before this PR, that the provided outputs were ignored when we get a cached kernel.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 9 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: We can also assert here when the provided output doesn't match.

File: torch/csrc/jit/codegen/cuda/fusion.cpp (hunk 10 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:617]: > I think the check here should be if it has the output, then it must have the input. Not that they should both always be here.

This API is used in parser. So at the time when we are defining the computation, segmentation hasn't happened yet. Hence the check here is just a sanity check. Because the only thing we can support via integration is to alias from an output to an input. I agree the other check is a lot more important, but this one also confines the scope of our aliasing support.

> add a comment that this is due to segmentation where we may use the input, and we could alias it, but if we don't produce the output then aliasing doesn't have to happen.

I'll put a comment as well as a TODO. Will work with Shiming to get this done.

> Since remove/add/output/input can be called multiple times on a fusion, we should validate at lowering that if an output in the alias map is there, then the corresponding input must be registered.

Sounds good. I'll add that check.

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:2235]: Just opened #896

File: torch/csrc/jit/codegen/cuda/fusion.cpp (hunk 10 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:617]: Opened #898 to track this

**Comment by jjsjann123** [comment]
I think I've addressed/tracked review comments.

**Review by csarofeen — APPROVED** [review]
